File: slpie/ui/app/core/lexicon.js
/* The words this reader uses, and the one function that applies them.
 *
 * `t("finding", {plural: true, title: true})` → "Findings", or "Risks" for a
 * tenant whose profile says so. Every user-visible noun the interface renders
 * from data goes through here, so adapting terminology is a lookup rather than
 * a fork.
 *
 * Two rules carried over from the kernel, because a split vocabulary is worse
 * than one wrong word:
 *
 * **A missing term renders as its key.** A render is not the place to fail. The
 * profile is where a bad key raises, because that is authored and checkable —
 * `slpie/context/lexicon.py` refuses it there.
 *
 * **A control's word is never renamed.** Severities, gap kinds, verdicts and
 * target states are protected server-side, so anything arriving here is already
 * either a name a context may choose or a word it may not. This module does not
 * re-check that; it could not, and duplicating the rule would give it somewhere
 * to be different.
 *
 * **`core/` imports nothing.** The default vocabulary is baked into
 * `data/client.js`, and this module cannot reach it — the browser follows the
 * kernel's ring rule one level down, and reaching upward for a constant is how
 * a tier stops being reusable. So `shell.js`, the composition root that may
 * import everything, seeds it at boot with `seed(LEXICON)`, and `load()`
 * refreshes it from `GET /api/lexicon` once the context is known.
 *
 * Before it is seeded every lookup returns its key. That is the same fallback
 * a missing term gets, so an unseeded module is ugly rather than broken.
 */

import { emit } from "./bus.js";

let words = {};
let profile = "default";

/** The word for a key. Falls back to the key, never throws. */
export function t(key, { plural = false, title = false } = {}) {
  const term = words[key];
  const word = term ? (plural ? term.plural : term.word) : key;
  return title ? word.charAt(0).toUpperCase() + word.slice(1) : word;
}

/** A term's one-sentence gloss, for a tooltip. Empty when nobody wrote one. */
export function gloss(key) {
  return (words[key] && words[key].gloss) || "";
}

/**
 * Resolve a label that may be a term reference.
 *
 * `"{finding.plural|title}"` → "Findings". A literal passes through untouched,
 * which is what lets a block carry either without the manifest declaring which
 * — most labels are literals and forcing every one through a term key would
 * mean inventing terms for "Window" and "Burst".
 */
export function label(text) {
  if (!text || !text.includes("{")) return text || "";
  return text.replace(/\{([^}]+)\}/g, (whole, body) => {
    const [reference, ...flags] = body.split("|");
    const [key, form] = reference.split(".");
    if (!words[key]) return whole;
    return t(key, {
      plural: form === "plural",
      title: flags.includes("title"),
    });
  });
}

export function current() {
  return profile;
}

/**
 * Install the baked default. Called once by `shell.js` before the first draw.
 *
 * Takes the compact shape `data/client.js` carries — `{key: {word, plural,
 * gloss}}` — rather than the API's envelope, because the two arrive from
 * different places and converting at the boundary is cheaper than teaching this
 * module both.
 */
export function seed(baked) {
  if (!baked) return;
  words = { ...baked };
}

/** Swap in a context's words. Announced, so drawn screens redraw. */
export function apply(body) {
  if (!body || !body.terms) return;
  words = {};
  for (const [key, term] of Object.entries(body.terms)) {
    words[key] = {
      word: term.word,
      plural: term.plural || term.word,
      gloss: term.gloss || "",
    };
  }
  profile = body.name || "default";
  emit("lexicon", profile);
}

/**
 * Fetch the lexicon for the current context.
 *
 * Failure is silent on purpose: the baked default is a complete, correct
 * vocabulary, and a console that refuses to render because it could not confirm
 * its own labels would be trading a cosmetic problem for an outage.
 */
export async function load(fetcher = fetch) {
  try {
    const response = await fetcher("/api/lexicon");
    if (!response.ok) return;
    apply(await response.json());
  } catch (error) {
    /* keep the baked words */
  }
}

File: slpie/context/lexicon.py
"""The same platform, in the reader's words.

A framework stays generalistic: it ships `Table`, `List`, `Item`, and every
product built on it wears the framework's vocabulary rather than its own. This
one goes the other way. The kernel knows what a thing *is*; what a given reader
*calls* it is a different fact, and keeping the two apart is what lets one
console read as a platform-engineering tool to one tenant and a compliance tool
to another without forking a screen.

**The default is derived, not authored.** Terms come from `slpie/domain/*.py` —
whose package docstring already describes itself as "the vocabulary every other
layer is written in" — and from the top-level package names under `slpie/`,
each glossed by the first line of its own docstring. So the platform cannot ship
a word its own code does not use, and a term that stops existing stops being
offered.

**A profile may rename the product. It may never rename a control.** The words
that carry policy — every `Severity`, every `GapKind`, every `Verdict`, every
target state, and the vocabulary of refusal — are protected, and `overlay`
refuses them rather than silently ignoring them. A tenant renaming *refused* to
*pending* is how a control becomes invisible, and it would be invisible to us
too. The protected set is derived from the enums themselves, so a severity added
next year is protected the day it is added rather than the day somebody
remembers to add it to a list.

Stdlib only, ring 0. Reads docstrings and enum members; no parser, no schema
library, no network.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass, replace
from pathlib import Path
from typing import Any, Iterable, Iterator, Mapping

from ..errors import SlpieError


class LexiconError(SlpieError):
    """A profile tried to rename something it may not, or a term nobody defines."""


#: Key prefixes a profile may never overlay. Each names a vocabulary that a
#: decision is read from rather than a noun the product happens to use.
#:
#: `refusal` is the one that is not an enum and is the most important of them.
#: The rest are derived from their enums below, so the set grows with the code.
PROTECTED_PREFIXES = ("severity.", "gap.", "verdict.", "target.", "refusal.")


@dataclass(frozen=True, slots=True)
class Term:
    """One word this context uses, and where the word came from."""

    key: str
    word: str
    plural: str = ""
    verb: str = ""
    gloss: str = ""
    source: str = ""            # the module it was derived from, or the profile

    def __post_init__(self) -> None:
        if not self.key or not self.word:
            raise LexiconError("a term needs both a key and a word")
        if not self.plural:
            object.__setattr__(self, "plural", _plural(self.word))

    @property
    def protected(self) -> bool:
        """Whether a profile is forbidden from replacing this word."""
        return self.key.startswith(PROTECTED_PREFIXES)

    def to_dict(self) -> dict[str, Any]:
        return {
            "key": self.key, "word": self.word, "plural": self.plural,
            "verb": self.verb, "gloss": self.gloss, "source": self.source,
            "protected": self.protected,
        }


def _plural(word: str) -> str:
    """English plurals, to the depth this actually needs.

    Deliberately not a library and deliberately not clever. The words in play
    are the product's own nouns — node, finding, dependency, policy — and a
    profile that wants something irregular supplies `plural` explicitly, which
    is a smaller surface than an inflection engine that is wrong about
    somebody's domain vocabulary in a way they cannot override.
    """
    if word.endswith(("s", "x", "z", "ch", "sh")):
        return f"{word}es"
    if word.endswith("y") and word[-2:-1] not in "aeiou":
        return f"{word[:-1]}ies"
    return f"{word}s"


class Lexicon:
    """Stable keys, and the words one context puts on them."""

    def __init__(self, terms: Iterable[Term] = (), *, name: str = "default") -> None:
        self.name = name
        self._terms: dict[str, Term] = {}
        for term in terms:
            self._terms[term.key] = term

    # -- access ----------------------------------------------------------

    @property
    def terms(self) -> tuple[Term, ...]:
        return tuple(self._terms[key] for key in sorted(self._terms))

    def get(self, key: str) -> Term | None:
        return self._terms.get(key)

    def word(self, key: str, *, plural: bool = False, title: bool = False) -> str:
        """The word for a key, or the key itself when nobody defines one.

        Falling back to the key rather than raising is deliberate: a missing
        term should show a slightly ugly label, never take a screen down. The
        *profile* is where a bad key fails loudly, because that is authored and
        checkable; a render is neither.
        """
        term = self._terms.get(key)
        if term is None:
            return key
        word = term.plural if plural else term.word
        return word[:1].upper() + word[1:] if title else word

    @property
    def protected(self) -> tuple[Term, ...]:
        return tuple(term for term in self.terms if term.protected)

    @property
    def renameable(self) -> tuple[Term, ...]:
        return tuple(term for term in self.terms if not term.protected)

    # -- overlay ---------------------------------------------------------

    def overlay(self, changes: Mapping[str, Mapping[str, str]], *, name: str) -> "Lexicon":
        """A new lexicon with a profile's words applied over these.

        Three refusals, each of which would otherwise be a silent failure:

        * an unknown key — a profile typo that renames nothing and looks like it
          worked;
        * a protected key — the whole point of the protected set;
        * an empty word — a term that renders as nothing.
        """
        merged = dict(self._terms)
        for key, fields in sorted(changes.items()):
            held = self._terms.get(key)
            if held is None:
                raise LexiconError(
                    f"{key!r} is not a term this platform defines; "
                    f"the closest are {', '.join(self._near(key)) or 'none'}"
                )
            if held.protected:
                raise LexiconError(
                    f"{key!r} carries a decision, not a name, and cannot be "
                    f"renamed. Renaming {held.word!r} would make a control "
                    f"unrecognisable to the operator reading it — and to us."
                )
            word = str(fields.get("word") or "").strip()
            if not word:
                raise LexiconError(f"{key!r} was given an empty word")
            merged[key] = Term(
                key=key, word=word,
                plural=str(fields.get("plural") or "").strip(),
                verb=str(fields.get("verb") or held.verb),
                gloss=str(fields.get("gloss") or held.gloss),
                source=f"profile:{name}",
            )
        return Lexicon(merged.values(), name=name)

    def _near(self, key: str) -> tuple[str, ...]:
        head = key.split(".", 1)[0]
        return tuple(
            item for item in sorted(self._terms)
            if item.startswith(head[:4]) or head in item
        )[:3]

    # -- projection ------------------------------------------------------

    def to_dict(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "digest": self.digest,
            "terms": {term.key: term.to_dict() for term in self.terms},
        }

    def words(self) -> dict[str, dict[str, str]]:
        """The compact form the browser consumes — key to word and plural."""
        return {
            term.key: {"word": term.word, "plural": term.plural,
                       "gloss": term.gloss}
            for term in self.terms
        }

    @property
    def digest(self) -> str:
        body = "\n".join(
            f"{term.key}\x1f{term.word}\x1f{term.plural}" for term in self.terms
        )
        return hashlib.blake2b(body.encode("utf-8"), digest_size=16).hexdigest()

    def __len__(self) -> int:
        return len(self._terms)

    def __iter__(self) -> Iterator[Term]:
        return iter(self.terms)

    def __contains__(self, key: object) -> bool:
        return key in self._terms


# --- the derived default ------------------------------------------------


def _docline(path: Path) -> str:
    """The first line of a module's docstring, without importing it."""
    import ast

    try:
        tree = ast.parse(path.read_text(encoding="utf-8"), filename=str(path))
    except (OSError, UnicodeDecodeError, SyntaxError):
        return ""
    doc = ast.get_docstring(tree) or ""
    head = doc.split("\n", 1)[0].strip()
    # "Nodes — the things an ecosystem is made of." → the half that explains.
    for separator in (" — ", " - ", ": "):
        if separator in head:
            return head.split(separator, 1)[1].strip()
    return head


def _protected_terms() -> list[Term]:
    """The vocabulary a decision is read from. Derived from the enums.

    Deriving rather than listing is what keeps the protection honest: a
    `Severity` added next year is protected the day it is added, not the day
    somebody remembers to add it to a list here.
    """
    from ..audit.judge import Verdict
    from ..binding.target import Target
    from ..domain.finding import GapKind, Severity

    terms: list[Term] = [
        Term("refusal.refused", "refused",
             gloss="A guard declined this, and said why.",
             source="slpie/ui/api.py"),
        Term("refusal.reason", "reason", gloss="Why it was declined.",
             source="slpie/rbac/engine.py"),
    ]
    for member in Severity:
        terms.append(Term(
            f"severity.{member.value}", member.value,
            gloss=f"Severity {member.value}.",
            source="slpie/domain/finding.py",
        ))
    for member in GapKind:
        terms.append(Term(
            f"gap.{member.value}", member.value.replace("_", " "),
            gloss="Something the platform could not see, and what it cost.",
            source="slpie/domain/finding.py",
        ))
    for member in Verdict:
        terms.append(Term(
            f"verdict.{member.value}", member.value,
            gloss=f"Verdict {member.value}.", source="slpie/audit/judge.py",
        ))
    for member in Target:
        terms.append(Term(
            f"target.{member.value}", member.value,
            gloss=f"The {member.value} binding.", source="slpie/binding/target.py",
        ))
    return terms


def default(root: Path | None = None) -> Lexicon:
    """The lexicon this platform ships, derived from its own code."""
    from .index import _repository

    base = root or _repository()
    terms: list[Term] = []
    seen: set[str] = set()

    domain = base / "slpie" / "domain"
    for path in sorted(domain.glob("*.py")) if domain.is_dir() else ():
        if path.name == "__init__.py":
            continue
        key = path.stem
        terms.append(Term(
            key=key, word=key, gloss=_docline(path),
            source=f"slpie/domain/{path.name}",
        ))
        seen.add(key)

    slpie = base / "slpie"
    for path in sorted(slpie.iterdir()) if slpie.is_dir() else ():
        init = path / "__init__.py"
        if not path.is_dir() or not init.is_file() or path.name.startswith(("_", ".")):
            continue
        key = path.name
        if key in seen:
            continue
        terms.append(Term(
            key=key, word=key,
            # A package named `artifacts`, `plugins` or `connectors` is already
            # the plural, and running it through `_plural` produced
            # "artifactses". Passing the word through as its own plural is
            # right for every such package here, and a profile that disagrees
            # says so explicitly — which it can, and the default cannot.
            plural=key if key.endswith("s") else "",
            gloss=_docline(init),
            source=f"slpie/{key}/__init__.py",
        ))
        seen.add(key)

    terms.extend(_protected_terms())
    return Lexicon(terms, name="default")
